## **Data Wrangling**

* Data wrangling, also known as data munging or data remediation, involves cleaning, transforming, and preparing raw data into a format suitable for analysis.

* We will typically go through the data wrangling process prior to conducting any data analysis in order to ensure your data is reliable and complete. This way, we can be confident that the insights we draw are accurate and valuable.

### Primary Goals

* Ensure data quality and consistency.
* Transform data into a usable format.
* Handle missing values, outliers, and inconsistencies.
* Integrate data from multiple sources.


### Tools and Technologies

* Python Libraries: Pandas, NumPy, SciPy, scikit-learn
* R Packages: dplyr, tidyr, data.table.
* ETL Tools: Apache NiFi, Talend, Alteryx.

P.S. We will be using only python libraries in this tutorial.

### Outcome

* A clean, well-structured dataset ready for analysis.


### Data Wrangling Steps

* There are six steps in the munging process:

    1. Discovery

    2. Structuring

    3. Cleaning

    4. Enriching

    5. Validation

    6. Publishing

<div>
    <img src="../assets/data_wrangling_steps.png" width=400>
</div>

Data Wrangling Process
<div>
    <img src="../assets/data_wrangling_process.png" width=500>
</div>

**Step 1: Discover**  
Initially, your focus is on understanding and exploring the data you’ve gathered. This involves identifying data sources, assessing data quality, and gaining insights into the structure and format of the data. Your goal is to establish a foundation for the subsequent data preparation steps by recognizing potential challenges and opportunities in the data.

**Step 2: Structure**  
In the data structuring step, you organize and format the raw data in a way that facilitates efficient analysis. The specific form your data will take depends on which analytical model you’re using, but structuring typically involves reshaping data, handling missing values, and converting data types. This ensures that the data is presented in a coherent and standardized manner, laying the groundwork for further manipulation and exploration.

**Step 3: Cleaning**  
Data cleansing is a crucial step to address inconsistencies, errors, and outliers within the dataset. This involves removing or correcting inaccurate data, handling duplicates, and addressing any anomalies that could impact the reliability of analyses. By cleaning the data, your focus is on enhancing data accuracy and to make sure the data is as error-free as possible.

**Step 4: Enriching**  
Enriching your data involves enhancing it with additional information to provide more context or depth. This can include merging datasets, extracting relevant features, or incorporating external data sources. The goal is to augment the original dataset, making it more comprehensive and valuable for analysis. If you do add data, be sure to structure and clean that new data.

**Step 5: Validating**  
Validation ensures the quality and reliability of your processed data. You’ll check for inconsistencies, verify data integrity, and confirm that the data adheres to predefined standards. Validation helps in building your confidence in the accuracy of the dataset and ensures that it meets the requirements for meaningful analysis.

**Step 6: Publishing**  
Now your curated and validated dataset is prepared for analysis or dissemination to business users. This involves documenting data lineage and the steps taken during the entire wrangling process, sharing metadata, and preparing the data for storage or integration into data science and analytics tools. Publishing facilitates collaboration and allows others to use the data for their analyses or decision-making processes.

## **Let's Learn By Doing**

### **Titanic Dataset**
* The Titanic Dataset is a dataset curated on the basis of the passengers on titanic, like their age, class, gender, etc to predict if they would have survived or not. While there was some element of luck involved in surviving, it seems some groups of people were more likely to survive than others.

**About Titanic Dataset**

The predefined columns are -

* Passenger Class (Pclass): This categorical variable indicates the ticket class of each passenger, ranging from 1st class (wealthiest) to 3rd class (lower socioeconomic status).

* Name: The names of passengers, providing insight into their identities.

* Sex: Gender of passengers, categorized as male or female.

* Age: Age of passengers, providing information about the demographic composition of the Titanic's passengers.

* SibSp: Number of siblings/spouses aboard the Titanic, offering insight into family relationships.

* Parch: Number of parents/children aboard the Titanic, indicating family size and composition.

* Ticket: Ticket number, providing additional information about passenger accommodations and fare details.

* Fare: Fare paid by each passenger, which can be indicative of their ticket class and economic status.

* Cabin: Cabin number or location, offering insights into passenger accommodations.

* Embarked: Port of embarkation (C = Cherbourg, Q = Queenstown, S = Southampton), providing information about passengers' embarkation points.

* Survived: This binary variable indicates whether a passenger survived the disaster (1) or not (0), serving as the primary outcome variable for analyses.

Dataset can be downloaded from [here](https://www.kaggle.com/c/titanic/data). We will be using only train.csv for this tutorial.

In [1]:
# Import the necessary libraries
import pandas as pd

# Supress warnings
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Load the dataset
df = pd.read_csv("../data/titanic_train.csv")

In [3]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
print(df.shape)

(891, 12)


There are 891 rows (data points) and 12 columns (features).

In [5]:
# Get the descriptive statistics of the dataset
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [6]:
# Get concise summary of a DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [7]:
# Let's check for missing values
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [8]:
# Let's check for duplicates
df.duplicated().sum()

0

There are no duplicates. If there are any duplicates, then we have to remove those duplicates or redundant rows. We can use drop_duplicates() method.

Curious about how many survived??

In [9]:
df["Survived"].value_counts()

Survived
0    549
1    342
Name: count, dtype: int64

### **Handling Missing Values**

* Various ways to hanlde missing values:
    *  Imputation using Mean or Median or Mode:  fill missing values with the mean or median or mode of the available data
    *  Imputation using Predictive Models:  build a predictive model to estimate missing values based on other features.
    *  Imputation using Group Statistics: fill missing values using the mean or median feature of groups that share similar characteristics.For example,for imputing age, you can use the average age of passengers with the same Pclass or Sex:
    * Dropping Missing Values (If the number of missing values is large)

There are lots of missing values in Cabin. So we will drop Cabin column and impute the missing values for Age and Embarked columns.

In [10]:
# Drop Cabin column
df = df.drop("Cabin", axis=1)

# Mean Imputation for Age
# Calculate the median age
median_age = df["Age"].median()

# Fill missing values with median age
df["Age"] = df["Age"].fillna(median_age)

# Mode Imputation for Embarked
# Calculate the mode of Embarked
mode_embarked = df["Embarked"].mode()[0]

# Fill missing values with mode of Embarked
df["Embarked"] = df["Embarked"].fillna(mode_embarked)

# Check for missing values
df.isnull().sum()


PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

### Did you notice there are some features in the dataset which are not numeric values?
* For proper data analysis, we must convert all the non-numeric values to numeric.

* Handling categorical and numeric values is a critical step in preparing data for analysis and modeling. 

* In any dataset, features can broadly be classified into two types: categorical and numeric.

    1. Categorical Values: 
        * These are variables that represent categories and can take on one of a limited, fixed number of possible values. 
        * Examples in the Titanic dataset include 'Sex', 'Embarked', and 'Pclass'.

    2. Numeric Values: 
        * These are variables that represent numerical values and can be discrete or continuous. 
        * Examples in the Titanic dataset include 'Age', 'Fare', and 'SibSp'.

**Types of Categorical Values**

* There are two primary categories of categorical data: 

    * Nominal data: 
        * Nominal data is used for labeling variables without any quantitative value. 
        * Nominal data cannot be ordered or ranked and are mutually exclusive.
        * Example: 
            * Gender: male, female
            * Color: Red, Blue, Green
            * Nationality: American, Canadian, Nepalese
    * Ordinal data: 
        * Ordinal data is a type of categorical data where the categories have a meaningful order or ranking. 
        * However, the intervals between the categories are not necessarily equal or known.
        * Example:
            * Rating Scale: Poor, Fair, Good, Very Good, Excellent
            * Education Level: High School, Bachelor's, Master's, Doctorate
            * Customer Satisfaction: Very Unsatisfied, Unsatisfied, Neutral, Satisfied, Very Satisfied

### **Encoding Categorical Values**

* Categorical values need to be converted into a numerical format before they can be used in machine learning models. This process is known as encoding.

**1. Label Encoding**
* Label encoding assigns a unique integer to each category. 
* This method can be used when the categorical variable is ordinal (i.e., has a meaningful order).
* Example:
<div>
    <img src="../assets/label_encoding.png">
</div>

* This approach assumes an ordering of the categories: "Never" (0) < "Rarely" (1) < "Most days" (2) < "Every day" (3)

**2. One-Hot Encoding**
* One-hot encoding creates new columns indicating the presence (or absence) of each possible value in the original data.  
* This method is suitable for nominal categorical variables (i.e., no intrinsic ordering).
* Example:
<div>
    <img src="../assets/one_hot_encoding.png">
</div>

* Here color is the categorical variable with three categories: "Red", "Yellow", and "Green" but they do not have any meaningful order.

* The corresponding one-hot encoding contains one column for each possible value, and one row for each row in the original dataset. Wherever the original value was "Red", we put a 1 in the "Red" column; if the original value was "Yellow", we put a 1 in the "Yellow" column, and so on.

* The major drawback of this approach is whenever the categorical variables have a large number of distinct values, several challenges arise, such as increased dimensionality, potential sparsity, and computational inefficiency. One way to solve this problem is **frequency encoding** i.e. replace each category with its frequency or count in the dataset.

Identify the nominal and ordinal variables in Titanic Dataset

In [11]:
column_names = df.columns
column_names

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Embarked'],
      dtype='object')

In [12]:
# Here we are lising out the categorical variables of object type only
# Actually Pclass, SibSp, Parch are also categorical variables as they represent counts of discrete entities
categorical_variables = df.dtypes[df.dtypes == "object"].index
categorical_variables

Index(['Name', 'Sex', 'Ticket', 'Embarked'], dtype='object')

Let's find the number of unique values for each categorical variables

In [13]:
for category in categorical_variables:
    print(f"Number of unique values in {category} = {len(df[category].unique())}")

Number of unique values in Name = 891
Number of unique values in Sex = 2
Number of unique values in Ticket = 681
Number of unique values in Embarked = 3


Since Name, and Ticket have large number of distinct values, one-hot-encoding will not be efficient. We could use Name column to extract the title of person like ["Miss", "Mr.", "Mrs."] and then encode them. But, for this tutorial we will drop those columns. 

In [14]:
df_clean = df.drop(["Name", "Ticket"], axis=1)

In [15]:
df_clean.head()

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0            1         0       3    male  22.0      1      0   7.2500        S
1            2         1       1  female  38.0      1      0  71.2833        C
2            3         1       3  female  26.0      0      0   7.9250        S
3            4         1       1  female  35.0      1      0  53.1000        S
4            5         0       3    male  35.0      0      0   8.0500        S

In [16]:
df_clean.shape

(891, 9)

In [17]:
categorical_variables = df_clean.dtypes[df_clean.dtypes == "object"].index
categorical_variables

Index(['Sex', 'Embarked'], dtype='object')

* Both "Sex" and "Embarked" are nominal categorical variables. So we will use one-hot-encoding.
* Pandas has get_dummies() method for one-hot-encoding.
* We can use sklearn library for label encoding and one-hot-encoding.
* Install sklearn:
    > pip install scikit-learn

In [18]:
df_clean_encoded = pd.get_dummies(df_clean, columns=categorical_variables, drop_first=True)
df_clean_encoded

PassengerId  Survived  Pclass   Age  SibSp  Parch     Fare  Sex_male  \
0              1         0       3  22.0      1      0   7.2500      True   
1              2         1       1  38.0      1      0  71.2833     False   
2              3         1       3  26.0      0      0   7.9250     False   
3              4         1       1  35.0      1      0  53.1000     False   
4              5         0       3  35.0      0      0   8.0500      True   
..           ...       ...     ...   ...    ...    ...      ...       ...   
886          887         0       2  27.0      0      0  13.0000      True   
887          888         1       1  19.0      0      0  30.0000     False   
888          889         0       3  28.0      1      2  23.4500     False   
889          890         1       1  26.0      0      0  30.0000      True   
890          891         0       3  32.0      0      0   7.7500      True   

     Embarked_Q  Embarked_S  
0         False        True  
1         False       False  
2         False        True  
3         False        True  
4         False        True  
..          ...         ...  
886       False        True  
887       False        True  
888       False        True  
889       False       False  
890        True       False  

[891 rows x 10 columns]

In [19]:
df_clean_encoded.shape

(891, 10)

Example to use sklearn label encoder. This is just for demonstration purpose. We have already used one-hot-encoding.

In [20]:
from sklearn.preprocessing import LabelEncoder


# Initialize the LabelEncoder
label_encoder = LabelEncoder()
df_copy = df_clean.copy()

# Fit and transform the 'Sex' column
df_copy['Sex'] = label_encoder.fit_transform(df_copy['Sex'])
print(df_copy.head())


   PassengerId  Survived  Pclass  Sex   Age  SibSp  Parch     Fare Embarked
0            1         0       3    1  22.0      1      0   7.2500        S
1            2         1       1    0  38.0      1      0  71.2833        C
2            3         1       3    0  26.0      0      0   7.9250        S
3            4         1       1    0  35.0      1      0  53.1000        S
4            5         0       3    1  35.0      0      0   8.0500        S


In [21]:
df_clean_encoded.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Age            0
SibSp          0
Parch          0
Fare           0
Sex_male       0
Embarked_Q     0
Embarked_S     0
dtype: int64

In [22]:
df_clean_encoded.head()

PassengerId  Survived  Pclass   Age  SibSp  Parch     Fare  Sex_male  \
0            1         0       3  22.0      1      0   7.2500      True   
1            2         1       1  38.0      1      0  71.2833     False   
2            3         1       3  26.0      0      0   7.9250     False   
3            4         1       1  35.0      1      0  53.1000     False   
4            5         0       3  35.0      0      0   8.0500      True   

   Embarked_Q  Embarked_S  
0       False        True  
1       False       False  
2       False        True  
3       False        True  
4       False        True

* We have completed the data wrangling i.e. 
    * dropped irrelevant features
    * handled missing values
    * perfrom encoding for categorical values

* Now the data is ready for further analysis. So we will save the data.

In [23]:
# Saving the cleaned dataset
df_clean_encoded.to_csv("../data/titanic_clean.csv", index=False)